File: chai2010-protorpc-intro/protorpc-hello-server.cc
// Copyright 2014 <chaishushan{AT}gmail.com>. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "./hello.pb/hello.pb.h" // HL

#include <google/protobuf/rpc/rpc_server.h>
#include <google/protobuf/rpc/rpc_client.h>

struct HelloService: public hello::HelloService { // HL
	virtual const ::google::protobuf::rpc::Error Hello(
		const ::service::HelloRequest* in, ::service::HelloResponse* out
	) {
		out->set_msg(in->msg());
		printf("HelloService.Hello: in:\"%s\", out:\"%s\"\n",
			in->msg().c_str(), out->msg().c_str()
		);
		return ::google::protobuf::rpc::Error::Nil(); // HL
	}
};

int main() {
	::google::protobuf::rpc::Server server;
	server.AddService(new HelloService, true); // HL
	server.BindAndServe(9527);// HL
	return 0;
}

File: chai2010-protorpc-intro/protorpc-hello-server.go
// Copyright 2014 <chaishushan{AT}gmail.com>. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import (
	"log" // OMIT
	// OMIT
	"code.google.com/p/goprotobuf/proto"
	hello "github.com/chai2010/talks/chai2010-protorpc-intro/hello.pb"  // HL
)

type myHello struct{}

func (p *myHello) Hello(in *hello.HelloRequest, out *hello.HelloResponse) error { // HL
	out.Msg = proto.String("myHello.Hello: " + in.GetMsg())
	log.Println("myHello.Hello:", in.GetMsg())
	return nil
}

func main() {
	go hello.ListenAndServeHelloService("tcp", ":9527", new(myHello)) // HL
	log.Println("Start HelloService on :9527 ...")
	<-make(chan bool)
}
